### ISSUE TITLE
Cannot rename table if part of a view

### ISSUE THREAD

**Issue by chrisjlocke** [sqlitebrowser/sqlitebrowser#1686]
Details for the issue
--------------------
This may be linked to #1650.
Tried to rename a table, but it failed as it tried to do something with the view, and failed.

![image](https://user-images.githubusercontent.com/7872218/50400913-b4161600-0782-11e9-99ef-32b1c8a65798.png)

#### What did you do?
Tried to modify a table.  In fact, just opening the 'Edit table definition' window and clicking 'OK' causes the error (which obviously tries to rebuild the table)

#### What did you expect to see?
No error

#### What did you see instead?
An error.

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [x] Windows: ( _version:_ _10__ )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- Nightly (24 Dec)
- [ ] 3.11.0-alpha1 or 3.11.0-beta*
- [ ] 3.10.1
- [ ] Other: ___

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [ ] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Heh Heh Heh.  Good catch @chrisjlocke. :smile:

**Comment by MKleusberg** [comment]
For me this just works :confused: I get the same SQL log as you but without any errors. Can you maybe double check which SQLite version is used in the nightlies?

**Comment by MKleusberg** [comment]
Ok, this version should definitely work :slightly_frowning_face: 

@chrisjlocke You said "In fact, just opening the 'Edit table definition' window and clicking 'OK' causes the error (which obviously tries to rebuild the table)". Can you check if this is still happening in the latest nightly build? I fixed this problem and the new code is similar to the code which should have prevented #1677 from ever happening. So it's interesting to see if the new code works as expected or if it has the same problem, i.e. if just clicking 'OK' rebuilds the table or not :smile:

**Comment by justinclift** [comment]
As a data point (related to my previous comment), it turns out SQLite 3.25.3 is safe too.  That release also has the patch for the Magellan vulnerability.

**Comment by MKleusberg** [comment]
Again, worth checking if this issue persists after the fix for #1725 :wink:

**Comment by chrisjlocke** [comment]
While renaming a field works, the error occurs if the primary key switch is turned off.  Can do a screencast if required.

Basic steps:
Create table with primary key.  Save table.  Modify table.  Remove primary key toggle for primary key field.  Click OK.  Receive below error.

![image](https://user-images.githubusercontent.com/7872218/52871423-ed934e00-3141-11e9-9f37-c6f61bad3f55.png)

Adding a field works.  **Removing** a field generates the same error.

**Comment by MKleusberg** [comment]
Yeah, maybe a screencast would help. I didn't manage to reproduce it yet :smile:

**Comment by chrisjlocke** [comment]
Yay!  I can play with the cheesy music again!  🎶

**Comment by chrisjlocke** [comment]
So 'view1' is simply a 'select * ffrom table11'.  Removing the primary key on table11 causes the error.

https://screencast-o-matic.com/watch/cqnobX39di

**Comment by MKleusberg** [comment]
Some observations:
1) Initially I tried this with SQLCipher 3.4.1 which worked fine. Now I'm using SQLite 3.26.0 and get the same error.
2) While playing around with this I could also trigger an error message which referred to a view which doesn't mention the modified table at all but is selecting from a non-existing table instead.

So there is definitely some weirdness going on here.

**Comment by MKleusberg** [comment]
Ok, I just followed the instructions from the [SQLite website](https://www.sqlite.org/lang_altertable.html#otheralter) step by step using the command line tool and... :drum: ... I got the exact same error message:

```
SQLite version 3.26.0 2018-12-12 11:57:35
Enter ".help" for usage hints.
sqlite> pragma foreign_keys=off;
sqlite> begin transaction;
sqlite> create table new_table11(Field1 INTEGER, Field2 INTEGER, Field3 INTEGER, Field4 INTEGER);
sqlite> insert into new_table11 select * from table11;
sqlite> drop table table11;
sqlite> alter table new_table11 rename to table11;
Error: error in view view1: no such table: main.table11
```

So maybe it's not an issue in our code after all?

**Comment by justinclift** [comment]
There's a new SQLite release (3.27.1) out that mentions fixing an ALTER TABLE bug (item 11) with views:

&nbsp; &nbsp; https://www.sqlite.org/releaselog/3_27_1.html

Maybe this is due to that?

**Comment by justinclift** [comment]
Hmmm, maybe point 17 in that same SQLite release notes list too.

**Comment by MKleusberg** [comment]
There is no ```WITH``` clause and no redundant ```UNIQUE``` clauses involved in our case but the release notes definitely make it sound like there are some remaining issues in the SQLite code.

**Comment by MKleusberg** [comment]
[This](https://www.mail-archive.com/sqlite-users@mailinglists.sqlite.org/msg113102.html) sounds like somebody else is having the same issue (he's doing it wrong initially but later writes he gets the same error when doing it the right way). The SQLite developers sound sceptical though :wink:

**Comment by MKleusberg** [comment]
Here is something interesting. In [this issue](https://www.sqlite.org/src/tktview?name=b41031ea2b) drh is telling us a bit about the improved ```ALTER TABLE``` functionality of SQLite >= 3.25.0:
> In version 3.25.0, the ALTER TABLE statement has been enhanced to run a test-parse of the whole schema after it alters the schema, to make sure that the edits it performed on the schema didn't break anything.

To me it sounds like this extra sanity check is what is causing the problem described here which also explains why it worked for older versions of SQLite.

In [this issue](https://www.sqlite.org/src/tktview?name=31c6e64ff9) drh describes a very similar issue to the one here. He is using a trigger instead of a view but besides that is is pretty much the same problem. Again some more explanations from the ticket:
> The DROP TABLE removes the table t1, which leaves a dangling reference to t1 in the trigger. Then during the ALTER TABLE, the trigger is reparsed, but the reparse fails due to the dangling reference.

Again, to me it sounds exactly like the issue we are having here. The second ticket is still open, so no fix there yet. Last modification was 2018-10-03. Seems like an unusual long time for the SQLite devs to fix a bug.

**Comment by justinclift** [comment]
Maybe lets ask them?

**Comment by karim** [comment]
So this issue is triggered when using SQLite >= 3.25.0? It works fine with older versions of SQLite?

**Comment by MKleusberg** [comment]
@karim I haven't done any conclusive testing yet. The SQLCipher version I tried is based on SQLite 3.15 - so there are many SQLite versions I haven't tried yet. 3.25.0 mainly comes into mind because of the improved ALTER TABLE support it added.

@justinclift I wanted to post a quick comment on that issue but turns out you're not allowed to do that as an anonymous user. I'll see if I get around to write to the mailing list soon-ish :smile:

**Comment by justinclift** [comment]
@MKleusberg From memory, @chrisjlocke is already subscribed to that mailing list.  Maybe Chris can do it? :smile:

**Comment by karim** [comment]
I'm simply asking if this is affecting some specific versions and we have no way of fixing so we can add them to the list known issues.

**Comment by chrisjlocke** [comment]
Have posted to the SQLite mailing list...

**Comment by chrisjlocke** [comment]
...to which Dan has replied.

"

ALTER TABLE ... RENAME TO ... commands fail in new versions of SQLite if  the schema contains views or triggers that refer to tables or columns  that do not exist. That's what is happening here. To restore the legacy  behaviour, run: 

    PRAGMA legacy_alter_table = 1;

https://sqlite.org/pragma.html#pragma_legacy_alter_table

Dan.

**Comment by justinclift** [comment]
Thanks @chrisjlocke. :smile:

**Comment by RyanAelen** [comment]
Encountering the same problem when executing an ALTER TABLE RENAME where an existing VIEW uses the newly named table.
- SQLite version 3.37.2
- Ubuntu 22.04.2 LTS, Release: 22.04

Origin/need: Migrating data from the old table structure to a new one, where the temp table gets renamed to the original table.

Implemented workaround: Dropped the affected VIEWS and remade them after migration.
